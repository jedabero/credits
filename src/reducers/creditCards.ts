import { createEntityAdapter, createSlice } from '@reduxjs/toolkit';
import { CreditCard } from '../models/CreditCard';
import type { RootState } from '../store';

export const cardsAdapter = createEntityAdapter<CreditCard>({
  selectId: ({ id }) => id,
  sortComparer: (a, b) => a.name.localeCompare(b.name),
});
const initialState = cardsAdapter.getInitialState();

const cardsSlice = createSlice({
  name: 'creditCards',
  initialState,
  reducers: {
    addCard: cardsAdapter.addOne,
    updateCard: cardsAdapter.updateOne,
    removeCard: cardsAdapter.removeOne,
  },
});

export default cardsSlice.reducer;

export const { addCard, updateCard, removeCard } = cardsSlice.actions;
export const {
  selectAll: selectAllCreditCards,
  selectById: selectCreditCardById,
  selectEntities: selectCreditCards,
  selectIds: selectCreditCardsIds,
  selectTotal: selectCreditCardTotal,
} = cardsAdapter.getSelectors<RootState>(({ creditCards }) => creditCards);
