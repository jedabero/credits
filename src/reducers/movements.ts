import {
  createEntityAdapter, createSlice, createDraftSafeSelector, EntityId,
} from '@reduxjs/toolkit';
import { Movement } from '../models/Movement';
import type { RootState } from '../store';

export const movementsAdapter = createEntityAdapter<Movement>({
  selectId: ({ id }) => id,
  sortComparer: (a, b) => a.authNumber.localeCompare(b.authNumber),
});
const initialState = movementsAdapter.getInitialState();

const cardsSlice = createSlice({
  name: 'movements',
  initialState,
  reducers: {
    addMovement: movementsAdapter.addOne,
    updateMovement: movementsAdapter.updateOne,
    removeMovement: movementsAdapter.removeOne,
  },
});

export default cardsSlice.reducer;

export const { addMovement, updateMovement, removeMovement } = cardsSlice.actions;
export const {
  selectAll: selectAllMovements,
  selectById: selectMovementById,
  selectEntities: selectMovements,
  selectIds: selectMovementsIds,
  selectTotal: selectMovementTotal,
} = movementsAdapter.getSelectors<RootState>(({ movements }) => movements);
export const selectMovementsIdsByCardId = (cardId: EntityId) => createDraftSafeSelector(
  ({ movements }: RootState) => movements,
  (movements) => movements.ids
    .filter((id) => movements.entities[id]?.cardId === cardId),
);
export const selectMovementsByCardId = (cardId: EntityId) => createDraftSafeSelector(
  ({ movements }: RootState) => movements,
  (movements) => movements.ids
    .filter((id) => movements.entities[id]?.cardId === cardId)
    .map((id) => movements.entities[id]),
);
