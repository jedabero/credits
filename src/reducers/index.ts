import { combineReducers } from '@reduxjs/toolkit';
import creditCards from './creditCards';

export default combineReducers({
  creditCards,
});
