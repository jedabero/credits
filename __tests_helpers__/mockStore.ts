// eslint-disable-next-line import/no-extraneous-dependencies
import configureMockStore from 'redux-mock-store';
import { RootState } from '../src/store';

export default configureMockStore<RootState>();
