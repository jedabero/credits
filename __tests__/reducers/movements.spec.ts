import mockStore from '../../__tests_helpers__/mockStore';
import { selectMovementsByCardId } from '../../src/reducers/movements';

describe('Movements reducer', () => {
  describe('Select Movements By Card Id', () => {
    test('Empty', () => {
      const store = mockStore({
        _persist: { version: 0, rehydrated: true },
        creditCards: {
          entities: {},
          ids: [],
        },
        movements: {
          entities: {},
          ids: [],
        },
      });
      const movements = selectMovementsByCardId('SOME ID')(store.getState());
      expect(movements).toMatchSnapshot();
    });
    test('Found', () => {
      const store = mockStore({
        _persist: { version: 0, rehydrated: true },
        creditCards: {
          entities: {},
          ids: [],
        },
        movements: {
          entities: {
            'SOME OTHER ID': {
              id: 'SOME OTHER ID',
              cardId: 'SOME ID',
              authNumber: '0',
              date: '',
              description: '',
              value: 1,
              monthlyRate: 0.01,
              yearlyRate: 0.01,
              quotas: 1,
            },
          },
          ids: ['SOME OTHER ID'],
        },
      });
      const movements = selectMovementsByCardId('SOME ID')(store.getState());
      expect(movements).toMatchSnapshot();
    });
  });
});
