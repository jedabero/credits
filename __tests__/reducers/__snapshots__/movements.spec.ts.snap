// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movements reducer Select Movements By Card Id Empty 1`] = `[]`;

exports[`Movements reducer Select Movements By Card Id Found 1`] = `
[
  {
    "authNumber": "0",
    "cardId": "SOME ID",
    "date": "",
    "description": "",
    "id": "SOME OTHER ID",
    "monthlyRate": 0.01,
    "quotas": 1,
    "value": 1,
    "yearlyRate": 0.01,
  },
]
`;
